### PR TITLE
Minor code improvements Ui/Diff and Rev and more

### PR DESCRIPTION
**Pull request by Klap-in**
Fixes a lot of warning from IDEA IntelliJ.

See also #3576

### CODE REVIEW

File: inc/Ui/Editor.php [hunk 1]
```diff
@@ -187,7 +188,7 @@ public function show()
      *
      * the default action for EDIT_FORM_ADDTEXTAREA
      *
-     * @param mixed[] $data
+     * @param array{wr: bool, media_manager: bool, target: string, intro_locale: string, form: Form} $data
```
Comment (Klap-in) [inline inc/Ui/Editor.php:191]: This phpdocs does also not work, but gives more insight in the entries of the array.
/** @var dokuwiki\Form\Form $data['form'] */ is also not recognized.

Comment (splitbrain) [reply inc/Ui/Editor.php:191]: How about this:

```suggestion
     * @param array $data {wr: bool, media_manager: bool, target: string, intro_locale: string, form: Form}
```

Comment (Klap-in) [reply inc/Ui/Editor.php:191]: Bit searching: It seems Phpstorm/IntelliJ support it, but it should probably be marked with `@psalm-param`. Sounds not so practical yet.

https://suckup.de/2020/02/modern-phpdoc-annotations/

The `array{…}` notation is already recognized as array, so the split you propose is not needed, but I will check.

Comment (Klap-in) [reply inc/Ui/Editor.php:191]: if `@param array{wr: bool, media_manager: bool, target: string, intro_locale: string, form: Form} $data` is used, it is already recognized as array parameter. But the entries are not default recognized.
If I install PHPStan plugin in IDEA IntelliJ the entire line is recognized. Is this worth to recommend the use of the PHPStan plugin?

Comment (splitbrain) [reply inc/Ui/Editor.php:191]: my suggestion was not meant to be recognized by IntelliJ but by Humans. For IntelliJ marking the parameter as an array is good enough IMHO. If the phpstan syntax still works with vanilla IntelliJ I'm happy to have it, otherwise I would prefer to stay at something that works out of the box.

Comment (Klap-in) [reply inc/Ui/Editor.php:191]: Clear. Array is recognized without the PHPstan plugin.

**Comment by Klap-in** [comment]
* 86216bf moves caching of the revision info of the current page (for eventually external edit info) from the constructor to the `getRevisionInfo()`
 * In 4106d4f I showed also the current revision info of external edits (and external creates, but these are never in the recent log of course). @ssahara or did you choose only external deletions because the point below?
 
More general point for discussion: 
 * the external edits/deletions are only listed in the Recent Changes (?do=recent) page if the revision before it is still in that log. In the list these revisions will show up with a new date and description, but they are not sorted. Sorting is difficult, because the Recent changes list is read per 20 (or whatever the setting is). Sorting of only these 20 is not so useful. Sorting of the whole log is bad for performance, as the log length can be set to a high number.

Is this fine for everybody? or are there ideas for alternatives?

recentchanges_externaledits

**Comment by ssahara** [comment]
**About Recent changes list and external edit**
* As far as the technical aspect, I agree with @Klap-in . From a user side,  when external deletion occurred, pagename link turns red which implicates the information of the page is no more valid or wrong. However, the edit summary may tell different information, for example "meeting starts 9:30AM", and visitors will be confused what is correct information. Therefore I thought the recent entry of externally deleted should be updated from last edit to avoid informing inconsistent message by indicating explicitly "removed external edit" occurred. When very old page has externally deleted, inconsistent message will not happen as far as the page does not appear in recent changes list. 
* For the simple external edit, last change date is not correct, but we could expect that the editor did not want to change posted date for minor edit. The externally created pages do not appear in the recent changes list.

**Comment by Klap-in** [comment]
Ok, I will revert 4106d4f. So only update external deletions in the Recent changes list.

**Comment by Klap-in** [comment]
Ok, I will revert 4106d4f. So only update external deletions in the Recent changes list.

**Comment by Klap-in** [comment]
@ssahara I saw I have added some conflicting changes here if compared with your branch.

I had a quick look in your changes, I will have later a look further. I saw nice things, but I have to test it yet.

btw, amazing how many corner cases there are for the diff/rev Ui parts!
